BUG: Fix memory leak of buffer-info cache due to relaxed strides

seberg opened the pull request:
When relaxed strides is active (and has an effect), we recalculate
the strides to export "clean" strides in the buffer interface.
(Python and probably some other exporters expect this, i.e. NumPy
has fully switched to and embraced relaxed strides, but the buffer
interface at large probably not.)

The place where "fixing" the strides occured however meant that
when the strides are fixed, the old, cached buffer-info was not
reused when it should have been reused.

This moves the "fixing" logic so that reuse will occur.  It leaves
one issue open in that an array shaped e.g. `(1, 10)` is both
C- and F-contiguous. Thus, if it is exported as C-contiguous and
then as F-contiguous, and then *again* as C-contiguous, this will
work, but the last export will compare to the export as an F-contig
buffer and thus still leak memory.

Address gh-16934 (but does leave a small hole)


---

@charris marking this as backport candidate. But, that code got some small changes, so it may be the diff does not apply (although I think it should).

seberg committed:
BUG: Fix memory leak of buffer-info cache due to relaxed strides

When relaxed strides is active (and has an effect), we recalculate
the strides to export "clean" strides in the buffer interface.
(Python and probably some other exporters expect this, i.e. NumPy
has fully switched to and embraced relaxed strides, but the buffer
interface at large probably not.)

The place where "fixing" the strides occured however meant that
when the strides are fixed, the old, cached buffer-info was not
reused when it should have been reused.

This moves the "fixing" logic so that reuse will occur.  It leaves
one issue open in that an array shaped e.g. `(1, 10)` is both
C- and F-contiguous. Thus, if it is exported as C-contiguous and
then as F-contiguous, and then *again* as C-contiguous, this will
work, but the last export will compare to the export as an F-contig
buffer and thus still leak memory.

Address gh-16934 (but does leave a small hole)

seberg commented on numpy/core/src/multiarray/buffer.c:
```diff
@@ -437,6 +437,7 @@ _buffer_format_string(PyArray_Descr *descr, _tmp_string_t *str,
 /* Additional per-array data required for providing the buffer interface */
 typedef struct {
     char *format;
+    int contiguity;
```
```suggestion
```

seberg commented on numpy/core/tests/test_multiarray.py:
```diff
@@ -7169,9 +7169,10 @@ def test_padded_struct_array(self):
+    @pytest.mark.valgrind_error(reason="leaks buffer info cache temporarily.")
+    def test_relaxed_strides(self, c=np.ones((1, 10, 10), dtype='i8')):
+        # Note: c defined as parameter so that it is persistent and leak
+        # checks will notice gh-16934 (buffer info cache leak).
```
I have confirmed that this slightly strange modification of the test does cause it to fail when using pytest-leaks.

eric-wieser commented on numpy/core/src/multiarray/buffer.c:
```diff
@@ -504,9 +504,37 @@ _buffer_info_new(PyObject *obj)
-                info->shape[k] = PyArray_DIMS(arr)[k];
-                info->strides[k] = PyArray_STRIDES(arr)[k];
+
+            /*
+             * When NPY_RELAXED_STRIDES_CHECKING is used, some buffer users
+             * may expect a contiguous buffer to have well formatted strides
+             * also when a dimension is 1, but we do not guarantee this
+             * internally. Thus, recalculate strides for contiguous arrays.
+             * (This is unnecessary, but has no effect in the case where
+             * NPY_RELAXED_STRIDES CHECKING is disabled.)
+             */
```
Do we have any evidence these users exist?

eric-wieser replied:
Also, should this code be conditional on NPY_RELAXED_STRIDES_CHECKING ?

seberg replied:
Happy to make it conditional. Without it, its basically a no-op currently.

As to evidence: There used to be issues with scipy, etc. when we switched NumPy over.  From the comment, I somewhat think Python _used_ to have issues around it (I doubt it still has with a brief look at the code).

It would be nice to just remove this dance completely.  But, I am unwilling to do this in 1.19, so I think we should probably put this in anyway, and then consider doing the additional step with a big release note.  The problem is, that relaxed strides was not a simple matter: We had to do multiple releases with it enabled during RC, but disabled during release to flush out (potential) issues.  We can do that again of course (set relaxed strides to debug mode during release), but it might be painful.

eric-wieser commented on numpy/core/src/multiarray/buffer.c:
```diff
@@ -504,9 +504,43 @@ _buffer_info_new(PyObject *obj)
+             * (This is unnecessary, but has no effect in the case where
+             * NPY_RELAXED_STRIDES CHECKING is disabled.)
+             */
+            if (PyArray_IS_C_CONTIGUOUS(arr) && !(
+                    f_contiguous && PyArray_IS_F_CONTIGUOUS(arr))) {
+                Py_ssize_t sd = PyArray_ITEMSIZE(arr);
+                for (k = info->ndim-1; k >= 0; --k) {
+                    info->shape[k] = PyArray_DIMS(arr)[k];
+                    info->strides[k] = sd;
+                    sd *= info->shape[k];
+                }
+            }
+            else if (PyArray_IS_F_CONTIGUOUS(arr)) {
+                Py_ssize_t sd = PyArray_ITEMSIZE(arr);
+                for (k = 0; k < info->ndim; ++k) {
+                    info->shape[k] = PyArray_DIMS(arr)[k];
+                    info->strides[k] = sd;
+                    sd *= info->shape[k];
+                }
+            }
```
Something for a follow-up - this is overly strict, if the user did not request a contiguous buffer at all we still construct these replacement strides.

seberg replied:
IIRC there was a reason for that. Maybe Cython may sometimes requested a buffer without specifying the flag, but then checking for it anyway later-on (potentially even raising an error). I somewhat expects cython caught up over time with our interpretation of relaxed strides, but I am not 100% sure.

peterbell10 commented on numpy/core/src/multiarray/buffer.c:
```diff
@@ -596,6 +630,13 @@ _buffer_get_info(PyObject *obj)
+             *       info had already been exported before (causing a memory
+             *       leak).
+             */
             if (_buffer_info_cmp(info, old_info) == 0) {
```
If a new buffer info is always constructed anyway, what's the benefit of having a cache at all? Also, it's a list but I only ever see the last item in the list referenced. Couldn't it be replaced with a single object per array? Then there is no memory growth at all.

seberg replied:
Yeah, it would be nice to skip part of that sometimes... and would be a small speed enhancement.  But, we have to reuse the old one, so that you do not leak memory and we need to hold on to it to be able to free it again.

The problem is that we can only *free* these when the array is deleted, which is the reason for your memory leak.  This is a fundamental issue around the buffer protocol, which actually does also provide the answer to the issue (by providing a `free` function).  But due to some backcompat issues, we can't just use that...

Times change, and the need for that backcompat probably goes down.  Maybe one more thing that would be nice to try out if we had a major release, I doubt that it would create a big issue in most cases.

seberg replied:
Sorry, the important part is that, even though nobody likes that, arrays are currently mutable both in shape/strides and dtype... So we can't just stick to a single exported buffer object :(.

peterbell10 replied:
So to sum up, it isn't a cache in the usual sense but instead a store of all buffer objects.

I see this was 577dbbd844b60958a59e062dae428a34a9c3644f which is 11 years old. If the fix 11 years old, perhaps that `PyArgs_ParseTuple` issue has been fixed in the mean time?

seberg replied:
No, that issue is still in `PyArg_ParseTuple` with an ugly comment as well there.  It may be that *users* of it are less now though.  And yes, you are right, the name "cache" is a misnomer, since it is persistent.

eric-wieser commented on numpy/core/src/multiarray/buffer.c:
```diff
@@ -504,9 +504,43 @@ _buffer_info_new(PyObject *obj)
+
+#if NPY_RELAXED_STRIDES_CHECKING
+            /*
+             * When NPY_RELAXED_STRIDES_CHECKING is used, some buffer users
+             * may expect a contiguous buffer to have well formatted strides
+             * also when a dimension is 1, but we do not guarantee this
+             * internally. Thus, recalculate strides for contiguous arrays.
+             * (This is unnecessary, but has no effect in the case where
+             * NPY_RELAXED_STRIDES CHECKING is disabled.)
+             */
```
```suggestion
             * Some buffer users who requested a contiguous buffer may expect
             * it to have well formatted strides even when a dimension is 1,
             * but we do not guarantee this internally when
             * NPY_RELAXED_STRIDES_CHECKING is used, so must recalculate
             * strides. For simplicity, we do this for all contiguous arrays,
             * even if contiguity was not explicitly requested.
             */
```

eric-wieser commented on numpy/core/src/multiarray/buffer.c:
```diff
@@ -596,6 +630,13 @@ _buffer_get_info(PyObject *obj)
             item = PyList_GetItem(item_list, PyList_GET_SIZE(item_list) - 1);
             old_info = (_buffer_info_t*)PyLong_AsVoidPtr(item);
 
+            /*
+             * TODO: In rare cases when an array is both C- and
+             *       F-contiguous, and buffers are exported alternating
+             *       we may always have a cache miss even though the correct
+             *       info had already been exported before (causing a memory
+             *       leak).
+             */
```
This seems very easy to fix - just make `key` the tuple `PyTuple_Pack(2, PyLong_FromVoidPtr((void*)obj), PyBool_FromLong(f_contiguous))` (with error checking).

seberg replied:
Hmmm, it does mean you have to delete two elements on every single array deallocation, though.

seberg replied:
So it trades of something that should never happen, with a small speed penalty on every deallocation right now.

seberg replied:
Sorry, I guess that is not true, it might happen, but its unlikely common. The other alternative is to always probe 2 buffers into the past for "reuse", and actually, I guess only when the array is both C- and F-contiguous, which is quick to check.

eric-wieser replied:
If you think using a `(pointer, order)` key is two expensive, then probably best to add that justification to a source comment.

seberg replied:
I think the issue was that I had neither of those ideas when writing that comment. And I do hate 98% solutions. I will have a quick look whether inspecting two list elements is easy, if not, I will expand the comment.

seberg replied:
OK, I pushed a fix, too often just staring at it and wondering takes more time than the fix :(. I have added a test, which trips if run with `pytest-leaks`. I have confirmed that it fails before and passes after.

eric-wieser commented on numpy/core/src/multiarray/buffer.c:
```diff
@@ -596,6 +630,13 @@ _buffer_get_info(PyObject *obj)
             item = PyList_GetItem(item_list, PyList_GET_SIZE(item_list) - 1);
             old_info = (_buffer_info_t*)PyLong_AsVoidPtr(item);
```
As an aside, I think we could get a reasonable speedup by using `Py_capsure` objects instead of `PyLong_AsVoidPtr` here - converting pointers to big integers and back seems silly when we could just keep the pointer around.

seberg replied:
Possibly (and probably nicer in any case).  I would prefer the approach in: gh-16938, but I don't have the mental capacity to figure out how annoying growing the struct is (I doubt its very) and mainly what could be done to reduce the chance of issues.

seberg committed:
BUG: Fix leak for relaxed strides when exporting both C- and F-order

Exporting these multiple times alternating would previously cause
a new buffer-info to be created each time.

charris commented:
@seberg @eric-wieser There is a push to release a 1.19.2, is this PR ready?

seberg commented:
Hmm, yeah, I think it is, I am not sure the last commit was ever reviewed though.

charris commented:
Looks like this is probably ready, but I'm going to push off to 1.19.3 for backport. We will need to make that release for python 3.9 and (hopefully) the Windows 10 version 2004 fix.

seberg committed:
MAINT: simplify buffer info reuse check

seberg commented:
I pushed a tiny bit of code cleanup in preparation of a fix for gh-17294.

eric-wieser commented on numpy/core/tests/test_multiarray.py:
```diff
@@ -7196,6 +7197,21 @@ def test_relaxed_strides(self):
 
+    @pytest.mark.valgrind_error(reason="leaks buffer info cache temporarily.")
+    @pytest.mark.skipif(not np.ones((10, 1), order="C").flags.f_contiguous,
+            reason="Test is unnecessary (but fails) without relaxed strides.")
```
why does this fail without relaxed strides?

seberg replied:
The array will be only C contiguous, so you can't request an f-contiguous buffer.

eric-wieser replied:
Makes sense, thanks. Am I right in thinking code-coverage skips this test, based on the review comment it made in the diff?

seberg replied:
That seems strange, unfortunately the comment is gone now due to the small comment change, so have to see which branch is uncovered and whether it should be later.

seberg replied:
Maybe the uncovered pass was within the `else` block of `#if NPY_RELAXED_STRIDES_CHECKING`? That can't be hit, since codecov should be running in the default setup.

seberg replied:
Well, the codecoverage didn't lie, the code got messed up in a cleanup try... The test was missing the strides check to notice it :(.

anyway, fixed now...

eric-wieser commented on numpy/core/src/multiarray/buffer.c:
```diff
@@ -800,7 +825,7 @@ void_getbuffer(PyObject *self, Py_buffer *view, int flags)
     }
 
     /* Fill in information */
```
```suggestion
    /* Fill in information: void scalars are always C-order (0) */
```

eric-wieser reviewed (commented):
Sorry for the delay reviewing this - I think the code looks good, just want to check I understand the test

charris commented:
What is the status of this? The CircleCI failures are fixed and master and can be ignored.

seberg committed:
Fixup, an earlier cleanup got things wrong and I forgot to test the strides :(

mattip commented:
Thanks @seberg
